kernel/CMakeLists.txt

```diff
@@ -52,7 +52,10 @@ target_sources_ifdef(
   userspace.c
   )
 
+get_property(E_KERNEL_ENTRY GLOBAL PROPERTY E_KERNEL_ENTRY)
 
 add_dependencies(kernel offsets_h)
 
-target_link_libraries(kernel zephyr_interface)
+target_link_libraries(kernel PUBLIC zephyr_interface)
```

Comment:
Changing the target_link_libraries signature of 'kernel' should be in it's own commit with an accompanying commit message that explains the reprecussions in case it breaks any code that does:

target_link_libraries(my_lib kernel)

that code needs to be ported to 

target_link_libraries(my_lib PUBLIC kernel)

I don't expect it to break anything, as I can't imagine a reason users would be linkin with the kernel, but it is best to be safe (and to separate changes),

Reply:
Why would it need to do that ?
From:
https://cmake.org/cmake/help/v3.12/command/target_link_libraries.html

> Library dependencies are transitive by default with this signature. When this target is linked into another target then the libraries linked to this target will appear on the link line for the other target too. This transitive “link interface” is stored in the INTERFACE_LINK_LIBRARIES target property and may be overridden by setting the property directly.

Reply:
Because of a bug in CMake that they made into a feature ;)

https://cmake.org/cmake/help/v3.0/policy/CMP0023.html

Reply:
Yes, but placing PUBLIC on the kernel library doesn't enforce that linking my_lib to kernel, as your line:

> target_link_libraries(my_lib kernel)

present needs to be modified.
target_link_libraries(my_lib kernel)
will still behave the same.

Reply:
Correct.

I am not trying to say that this will transitively affect all other libraries.

I am trying to say that if someone has been doing

target_link_libraries(kernel some_lib)

then they will get an error until they port their code to look like this:

target_link_libraries(kernel PUBLIC some_lib).

Hmm, on second thought, I guess this is safe, because you can only modify kernel from the same CMakeLists.txt file (directory) that created it, which is zephyr/CMakeLists.txt.

I guess we are lucky we are using 'add_subdirectory(kernel)' and not 'include(kernel)' in this case ;)

Reply:
ok, now your comment makes much more sense.
and this is exactly why CMakeLists.txt files should be used in COMBINATION with include.
As I show in #8451 

Libraries/subsystems, such as kernel, net, bluetooth, etc. should still be included using add_subdirectory, but INSIDE such library, the use of include([\<subdir\>/]\<file\>.cmake) can give a much better handling of the lib.

So the use of add_subdirectory in this case is not luck. 

And btw. the library itself can still be modified, i.e.
you can do:
target_sources(kernel PRIVATE <somefile.c>) anywhere in the CMake tree, regardless of whether you use include or add_subdirectory.
add_subdirectory prevents only the use of:
target_link_libraries(kernel ...)
outside of the CMakeLists.txt directory.